Allow connecting signals to existing methods without opening the script editor

passivestar opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/8351

Adds an editor setting to prevent the script editor from opening when connecting a signal to an existing method

<img width="933" alt="Screenshot 2024-03-08 at 03 40 27" src="https://github.com/godotengine/godot/assets/60579014/c519c09f-7ce5-44ff-9e8b-a0404c02a3ad">

Note that script editor opening when connecting to an existing method [was a bug](https://github.com/godotengine/godot-proposals/issues/8351#issuecomment-1893996727), but we're preserving that behavior by default

_EDIT: Switched from doing this by holding Ctrl when pressing Connect to an editor setting_

passivestar committed:
Allow connecting signals to existing methods without opening the script editor

Mickeon commented:
While I do find this to be a decent feature, the editor has a discoverability issue for these kinds of things and this additional CTRL shortcut does not help.

At least, if possible, what I would do is have the text on the change **completely** when holding the CTRL key. I've never seen something like this done by the editor before, but _at least_ it makes the feature more noticeable.

In my opinion though, this could just as easily be an Editor Setting, or perhaps, a toggle on the Connect dialog itself, or both?

Mickeon commented on scene/gui/dialogs.cpp:
```diff
@@ -195,6 +195,14 @@ String AcceptDialog::get_ok_button_text() const {
 	return ok_button->get_text();
 }
 
+void AcceptDialog::set_ok_button_tooltip_text(String p_ok_button_tooltip_text) {
+	ok_button->set_tooltip_text(p_ok_button_tooltip_text);
+}
+
+String AcceptDialog::get_ok_button_tooltip_text() const {
+	return ok_button->get_tooltip_text();
+}
+
```
Raising eyebrows, these new methods feel a little _too_ niche. Should not exist unless other things in the engine would greatly benefit from them.

passivestar replied:
> get_ok_button() already exists, so these methods are not needed.

ah nice

passivestar commented on editor/connections_dialog.cpp:
```diff
@@ -861,6 +861,10 @@ ConnectDialog::ConnectDialog() {
 	error->set_ok_button_text(TTR("Close"));
 	set_ok_button_text(TTR("Connect"));
+
+	bool macos = OS::get_singleton()->has_feature("macos") || OS::get_singleton()->has_feature("web_macos") || OS::get_singleton()->has_feature("web_ios");
+	String keycode_name = macos ? find_keycode_name(Key::META) : find_keycode_name(Key::CTRL);
```
true

passivestar commented:
> At least, if possible, what I would do is have the text on the change completely when holding the CTRL key. I've never seen something like this done by the editor before, but at least it makes the feature more noticeable.

I've never seen anything like that either and I'm not sure it would help discoverability if you need to hold CTRL to discover it

> In my opinion though, this could just as easily be an Editor Setting, or perhaps, a toggle on the Connect dialog itself, or both?

I'm 100% on board to make it an editor setting (with default being the current behavior) and nothing else. In my opinion it's all that's really needed here, especially considering that editor opening on existing methods was unintended

KoBeWi commented:
Seems like it breaks connecting to native methods:

https://github.com/godotengine/godot/assets/2223172/35b9937e-c4d5-4d4a-8aa1-d7e5e77016b3

I held Ctrl, but not only it opened the editor, it redefined `queue_free()`.

passivestar commented:
> Seems like it breaks connecting to native methods

Fixed that and also switched from the modifier key to an editor setting

Here's how it works now:

1. If the method exists in an inherited class, either native or script, it creates a connection without adding anything or opening the editor (as it did before)
2. If the method exists in the script on the target node itself it checks the editor setting to decide whether to open the editor or not

I put it into **"Text Editor -> Behavior -> Navigation"** (let me know if there's a better place)

<img width="933" alt="Screenshot 2024-03-08 at 03 40 27" src="https://github.com/godotengine/godot/assets/60579014/c519c09f-7ce5-44ff-9e8b-a0404c02a3ad">

KoBeWi commented on doc/classes/EditorSettings.xml:
```diff
@@ -1011,6 +1011,9 @@
 			If [code]true[/code], the caret will be moved when right-clicking somewhere in the script editor (like when left-clicking or middle-clicking). If [code]false[/code], the caret will only be moved when left-clicking or middle-clicking somewhere.
 		</member>
+		<member name="text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method" type="bool" setter="" getter="">
+			If [code]true[/code], opens the script editor when connecting a signal to an existing method from the Node dock.
```
```suggestion
			If [code]true[/code], opens the script editor when connecting a signal to an existing script method from the Node dock.
```
More precise.

KoBeWi commented on editor/connections_dialog.cpp:
```diff
@@ -906,25 +906,34 @@ void ConnectionsDock::_make_or_edit_connection() {
+		// Check in target's own script.
+		int line = scr->get_language()->find_function(cd.method, scr->get_source_code());
+		if (line != -1) {
+			add_script_function_request = bool(EDITOR_GET("text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method"));
```
```suggestion
			add_script_function_request = EDITOR_GET("text_editor/behavior/navigation/open_script_when_connecting_signal_to_existing_method");
```
No need to be explicit when assigning.

akien-mga commented:
Thanks!
